packages/dds/tree/src/feature-libraries/default-schema/schemaChecker.ts

```diff
@@ -0,0 +1,105 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	MapTree,
+	StoredSchemaCollection,
+	TreeFieldStoredSchema,
+	TreeNodeStoredSchema,
+	LeafNodeStoredSchema,
+	ObjectNodeStoredSchema,
+	MapNodeStoredSchema,
+} from "../../core/index.js";
+import { FullSchemaPolicy } from "../modular-schema/index.js";
+import { Multiplicity } from "../multiplicity.js";
+import { allowsValue } from "../valueUtilities.js";
+
+export function isNodeInSchema(
+	node: MapTree,
+	schema: TreeNodeStoredSchema,
+	nodeSchemaCollection: StoredSchemaCollection,
+	schemaPolicy: FullSchemaPolicy,
+): boolean {
+	if (
+		schema instanceof LeafNodeStoredSchema &&
+		(node.value === undefined ||
+			node.fields.size !== 0 ||
+			!allowsValue(schema.leafValue, node.value))
+	) {
+		return false;
+	}
+
+	if (schema instanceof ObjectNodeStoredSchema) {
+		if (node.fields.size !== schema.objectNodeFields.size) {
+			return false;
+		}
+		for (const [fieldKey, field] of node.fields) {
+			const fieldSchema = schema.objectNodeFields.get(fieldKey);
+			if (
+				fieldSchema === undefined ||
+				!isFieldInSchema(field, fieldSchema, nodeSchemaCollection, schemaPolicy)
+			) {
+				return false;
+			}
+		}
+	}
+
+	if (schema instanceof MapNodeStoredSchema) {
+		for (const field of node.fields.values()) {
+			if (!isFieldInSchema(field, schema.mapFields, nodeSchemaCollection, schemaPolicy)) {
+				return false;
+			}
+		}
+	}
+
+	return true;
```

Comment:
The way this is written, its harder than necessary to tell if its correct.

I think structuring the whole thing in the style below would make it easier to analyze, and provide a nice error if you miss a case (instead of returning true).


```
if (schema instanceof LeafNodeStoredSchema) {
....
} else if (schema instanceof ...) {
}
...
else {
assert("unknown TreeNodeStoredSchema type")
}
```
}

Reply:
returning true is not for when a case is missed, is for asserting the node match the schema, but the style you suggest makes sense and allow not returning true when a case is missed, so I implemented it.